sdks/java/extensions/sql/datacatalog/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/datacatalog/DataCatalogTableProvider.java

```diff
@@ -138,8 +143,41 @@ private Table loadTableFromDC(String tableName) {
     }
   }
 
-  @Override
-  public BeamSqlTable buildBeamSqlTable(Table table) {
-    return delegateProviders.get(table.getType()).buildBeamSqlTable(table);
+  private static DataCatalogBlockingStub createDataCatalogClient(
+      DataCatalogPipelineOptions options) {
+    return DataCatalogGrpc.newBlockingStub(
+            ManagedChannelBuilder.forTarget(options.getDataCatalogEndpoint()).build())
+        .withCallCredentials(
+            MoreCallCredentials.from(options.as(GcpOptions.class).getGcpCredential()));
+  }
+
+  private static Map<String, TableProvider> getSupportedProviders() {
+    return Stream.of(
+            new PubsubJsonTableProvider(), new BigQueryTableProvider(), new TextTableProvider())
+        .collect(toMap(TableProvider::getTableType, p -> p));
+  }
+
+  private Table toCalciteTable(String tableName, Entry entry) {
+    if (entry.getSchema().getColumnsCount() == 0) {
+      throw new UnsupportedOperationException(
+          "Entry doesn't have a schema. Please attach a schema to '"
+              + tableName
+              + "' in Data Catalog: "
+              + entry.toString());
+    }
+    Schema schema = SchemaUtils.fromDataCatalog(entry.getSchema());
+
+    Optional<Table.Builder> tableBuilder = tableFactory.tableBuilder(entry);
+    if (tableBuilder.isPresent()) {
+      return tableBuilder.get().schema(schema).name(tableName).build();
```

Comment:
nit: This pattern gives you all the pain of `Optional` with none of the benefits.

Reply:
What should I do instead?

Reply:
FWIW this is the only approved use of `Optional`. Since it isn't serializable, it isn't "data" per se, and for optional arguments they encourage overrides or different method names. It is specifically for methods that may or may not return a thing.

But I take your point about chaining things that each may or may not handle the input. I don't love it. I wish they could just be keyed on some enum in the proto. Just not the current status of the Data Catalog API as I understand it.